### ISSUE TITLE
Can not get accurate tracking

### ISSUE THREAD

**Issue by ZiyiZhang0912** [DeepLabCut/DeepLabCut#1107]
OS: Win 10
DeepLabCut Version: 2.2b8
Anaconda env used: DLC-GPU (cloned from Alex's github)
Tensorflow version: many, installed with pip (see below)
Cuda version: 10

First of all, I would like to thank the function of deeplabcut for the efficiency brought to my work! 

I cannot get accurate tracking every time in the video tracking.

For the first time, I used six videos from three different angles with an average of one minute to train the network. The sizes of these videos are all 1G-3G. In training, I used kmeans to extract 40 frames for each video, that is, a total of 240 frames were used for training, training until the LOSS was 0.002. The remaining parameters are default parameters. Using the network this time to track my animal behavior videos, the results obtained are not perfect, but accurate enough.

This is the result obtained after evaluating_network.

<img width="952" alt="8b4bfe3eff76b8296438fd890f11afc" src="https://user-images.githubusercontent.com/62039483/107141060-50023c00-6961-11eb-9aff-ffda580dd958.png">

However, when I tried to train the network for the second time, I replaced the original video with 8 videos with an average of 4 minutes and shot from four different angles. The size of these videos is 300M. In training, I still use kmeans to extract 40 frames for each video, that is, a total of 320 frames are used for training, and training until the LOSS is 0.002 (that is, repeat the steps of the first training). However, the visual tracking results of this network are very poor, with almost no accurate tracking points.

Also attach the results obtained after evaluating_network.

<img width="759" alt="7f1ecd70d7d8c024ba63ca208d24db1" src="https://user-images.githubusercontent.com/62039483/107141070-5ee8ee80-6961-11eb-8c35-c8c20fd9e72d.png">

The error of the test is very large, but I am not sure why.

I am a novice, and I have not been able to get the reason for the huge gap between the two training results. Is it related to the  format and size of the video? So I want to ask questions here to get answers that can help me! thank you very much.

**Comment by KonradDanielewski** [comment]
I'm confused by the number of frames you actually extracted and labeled. If you extract 40 frames per video and first project has 6 videos, that'd be 240 frames and for the second with 8 videos that'd be 320 frames, yet you say you labeled 120 frames for each project. Both test errors are quite big and I'd assume that is because your dataset is small. 120 frames for 3 or 4 different angles (depending on how different they are) is way to little to train a robust model. Train error will be small after proper training because the model get's good at recognizing what it has seen. The problem rises when it hasn't seen enough differently looking frames so it can't generalize properly (recognize bpts on frames it hasn't seen). Both of your test-error results are probably not good enough for any further analysis (13 px error when taking only likelihood above 0.9 is not great, it should be closer to 3-4 px at most).

**Comment by ZiyiZhang0912** [comment]
> I'm confused by the number of frames you actually extracted and labeled. If you extract 40 frames per video and first project has 6 videos, that'd be 240 frames and for the second with 8 videos that'd be 320 frames, yet you say you labeled 120 frames for each project. Both test errors are quite big and I'd assume that is because your dataset is small. 120 frames for 3 or 4 different angles (depending on how different they are) is way to little to train a robust model. Train error will be small after proper training because the model get's good at recognizing what it has seen. The problem rises when it hasn't seen enough differently looking frames so it can't generalize properly (recognize bpts on frames it hasn't seen). Both of your test-error results are probably not good enough for any further analysis (13 px error when taking only likelihood above 0.9 is not great, it should be closer to 3-4 px at most).

I am very sorry for my mistake, it is indeed 240 frames and 320 frames respectively. Do you think 240 frames and 320 frames are not enough for three perspectives and four perspectives? And my cameras' setup, the angle between every two cameras is 90 degrees, do you think that this angle of views will also make the network performance poor?

**Comment by KonradDanielewski** [comment]
It's hard to say what might be the issue. For so many big camera angles I'd go for more frames (by extracting outliers from videos analyzed with the model you have). Also having more labels might be helpful as far as I know (how many do you have?) Could you provide the plots from a video you analyzed using the model? Or plots for all the differently angled videos.

**Comment by ZiyiZhang0912** [comment]
> It's hard to say what might be the issue. For so many big camera angles I'd go for more frames (by extracting outliers from videos analyzed with the model you have). Also having more labels might be helpful as far as I know (how many do you have?) Could you provide the plots from a video you analyzed using the model? Or plots for all the differently angled videos.




> It's hard to say what might be the issue. For so many big camera angles I'd go for more frames (by extracting outliers from videos analyzed with the model you have). Also having more labels might be helpful as far as I know (how many do you have?) Could you provide the plots from a video you analyzed using the model? Or plots for all the differently angled videos.

I am sorry for not responding in time, because I experienced a vacation.

The following is the result of using the plot_trajectories function after analyzing a 30-second video. It seems that the network is performing badly.

![hist](https://user-images.githubusercontent.com/62039483/108447268-8bbecf00-729a-11eb-9f98-fba9eca80cf5.png)
![plot](https://user-images.githubusercontent.com/62039483/108447269-8ceffc00-729a-11eb-94dd-1551b6c3381a.png)
![plot-likelihood](https://user-images.githubusercontent.com/62039483/108447272-8e212900-729a-11eb-8d0d-5753de7f1800.png)

In addition, the video I used is exactly four different angle videos of four different individuals, totaling 16 videos, each of which is four minutes long. Do you think that when I select frames, I should extract more frames for an individual (for example, extract 100-200 frames for an individual), or should I select multiple bodies but a few frames for each individual?

Secondly, I have a body label of the back, but I have a front camera and a back camera. Do you think that for the back tracking of an individual, should I only mark the back of the individual, or treat the back and abdomen as the same Tracking points to label? (I don’t know if what I said is clear)

Thank you again for any replies and suggestions. It will be very helpful to me. Thank you!

**Comment by KonradDanielewski** [comment]
> In addition, the video I used is exactly four different angle videos of four different individuals, totaling 16 videos, each of which is four minutes long. Do you think that when I select frames, I should extract more frames for an individual (for example, extract 100-200 frames for an individual), or should I select multiple bodies but a few frames for each individual?
> 
> Secondly, I have a body label of the back, but I have a front camera and a back camera. Do you think that for the back tracking of an individual, should I only mark the back of the individual, or treat the back and abdomen as the same Tracking points to label? (I don’t know if what I said is clear)
> 
> Thank you again for any replies and suggestions. It will be very helpful to me. Thank you!

Well, the plots for one, show that there is a lot of missing detections of some bodyparts (which is actually a low likelihood detection but in reality the label would be nowhere near where it should be). Do bodyparts disappear a lot - meaning they are not visible to you? If yes, then this is both good and bad news - depends on your use case.  Since the angles are big, I'd go for closer to 100 labeled frames per camera angle (at least, realisticaly more like 150 per video). Not knowing how your videos really look and without knowing the whole use case the best advice I can give is to label far more frames and retrain the model (you can extract outliers and refine those to expand the dataset). Lastly I wouldn't use the same label for the back and abdomen. 

After retraining the plots should get a little bit better but if you have too many oclussions they will not change that much. If you need to have the detections every frame cause you're modelling something/doing 3D reconstruction using the pose estimation then interpolation and/or some form of spatiotemporal constrained filtering is probably your best bet since using data with missing values will not work.

So, first thing, label more and retrain the model and check it's performance again. Make the plots, maybe create labeled video for each camera angle to visually inspect the performance and get an idea what's going on.

**Comment by ZiyiZhang0912** [comment]
> Well, the plots for one, show that there is a lot of missing detections of some bodyparts (which is actually a low likelihood detection but in reality the label would be nowhere near where it should be). Do bodyparts disappear a lot - meaning they are not visible to you? If yes, then this is both good and bad news - depends on your use case. Since the angles are big, I'd go for closer to 100 labeled frames per camera angle (at least, realisticaly more like 150 per video). Not knowing how your videos really look and without knowing the whole use case the best advice I can give is to label far more frames and retrain the model (you can extract outliers and refine those to expand the dataset). Lastly I wouldn't use the same label for the back and abdomen.
> 
> After retraining the plots should get a little bit better but if you have too many oclussions they will not change that much. If you need to have the detections every frame cause you're modelling something/doing 3D reconstruction using the pose estimation then interpolation and/or some form of spatiotemporal constrained filtering is probably your best bet since using data with missing values will not work.
> 
> So, first thing, label more and retrain the model and check it's performance again. Make the plots, maybe create labeled video for each camera angle to visually inspect the performance and get an idea what's going on.

Yes, your guess is correct. There are many occlusion problems in my video, and the extracted frames also have motion blur, so there is inaccuracy when manually labeling. But because my purpose is to perform 3D tracking and there are space limitations, I have to set up four cameras with different orientations. For example, in the video taken by the front camera, the back half of the animal will be completely blocked.

What I am most uncertain of is, for example, when I label the elbow, there are multiple sides of the elbow in the actual video. Should I label the back, side and front of the elbow with the same label? This problem is similar to the back and the abdomen, but it is very different, because the back and the abdomen are the front and back, but the elbow is a 360-degree part.

Yes, I am trying to refine label and extract outlier label, hoping to improve. But is the quality of the video itself, perhaps the cause that most affects network performance?

For occlusion and motion blur, do you have any suggestions for improvement?

**Comment by KonradDanielewski** [comment]
Maybe some advice on 3D reconstruction, it will answer your question and maybe help your implementation:

1. Motion blur - if you have control over shutter speed set it higher (may also be called exposure time in your camera settings); ~8ms should be ok for rats, mice might need like 6ms but have yet to test this. For high shutter speed you have to provide very good lighiting conditions (this may not be that necessarily true but I need to do some testing on going with high gain and post-hoc video denoising when scene light is low) 

2. Occlusions - for good 3D reconstruction pairs of cameras have to see the same thing at the same time. This means that ideally from all angles you should be able to label all your bodyparts. Can you work out a setup that will allow you to place cameras in a way that will provide all 4 of them most visibility of the same bodyparts most of the time?

3. In kind of whole animal tracking I'm doing (and I assume you too) in case of small animals like rats/mice I'd consider small bodyparts like elbow to be the same thing from all angles and label it as such but back and abdomen are not only completely different and surrounded by different things, they are very far from eachother.

4. Probably most important, very good, robust model that gets accurate detections on all camera angles with best precission possible. This takes time and a lot of retraining, usually it's just an ongoing process of forever refining your model with every new experiment you recorded.

Lastly, when you've done what's mentioned above the best way it was possible in your conditions and still have some missing data don't hesitate to filter it and interpolate what's missing. You can interpolate 2D data upon which the reconstruction will be based, consider spatial and temporal constraints, smoothing of the jittering etc. Since you're bound to walk into werid things with triangulation, like your labels flying off to the edges of video etc. these are actually necesarry and good 3D reconstruction wouldn't be possible without them.

Hope this helps

**Comment by ZiyiZhang0912** [comment]
> Maybe some advice on 3D reconstruction, it will answer your question and maybe help your implementation:
> 
> 1. Motion blur - if you have control over shutter speed set it higher (may also be called exposure time in your camera settings); ~8ms should be ok for rats, mice might need like 6ms but have yet to test this. For high shutter speed you have to provide very good lighiting conditions (this may not be that necessarily true but I need to do some testing on going with high gain and post-hoc video denoising when scene light is low)
> 2. Occlusions - for good 3D reconstruction pairs of cameras have to see the same thing at the same time. This means that ideally from all angles you should be able to label all your bodyparts. Can you work out a setup that will allow you to place cameras in a way that will provide all 4 of them most visibility of the same bodyparts most of the time?
> 3. In kind of whole animal tracking I'm doing (and I assume you too) in case of small animals like rats/mice I'd consider small bodyparts like elbow to be the same thing from all angles and label it as such but back and abdomen are not only completely different and surrounded by different things, they are very far from eachother.
> 4. Probably most important, very good, robust model that gets accurate detections on all camera angles with best precission possible. This takes time and a lot of retraining, usually it's just an ongoing process of forever refining your model with every new experiment you recorded.
> 
> Lastly, when you've done what's mentioned above the best way it was possible in your conditions and still have some missing data don't hesitate to filter it and interpolate what's missing. You can interpolate 2D data upon which the reconstruction will be based, consider spatial and temporal constraints, smoothing of the jittering etc. Since you're bound to walk into werid things with triangulation, like your labels flying off to the edges of video etc. these are actually necesarry and good 3D reconstruction wouldn't be possible without them.
> 
> Hope this helps

Thank you very much for your suggestions and patient guidance, each one is helpful and enlightening to me.

Before this, I didn't know much about camera skills. I have learned how to adjust the shutter speed to correct motion blur. But for my experimental site, lighting is also something to be concerned about. I am working on this, and the experimental environment for 3D tracking is still under construction.

At present, I can only try to find the optimal strategy by using videos that are not ideal. This is the evaluation result I got after refine_label on one of the videos:

<img width="563" alt="76ccd9589e14513cf4b2ec9b1bff8d2" src="https://user-images.githubusercontent.com/62039483/108807548-b8971d00-75df-11eb-8acc-c958235ea6a1.png">

![hist](https://user-images.githubusercontent.com/62039483/108808466-ae761e00-75e1-11eb-832d-04fc248bf06b.png)
![plot](https://user-images.githubusercontent.com/62039483/108808478-b0d87800-75e1-11eb-8bf0-de9842fb6692.png)
![plot-likelihood](https://user-images.githubusercontent.com/62039483/108808483-b209a500-75e1-11eb-9b13-7993ac135059.png)

Although the test error is very small, it is even similar to the train error. However, the actual network performance is still not good, with only a slight improvement compared to before. I don’t understand why the test error has reached an acceptable value, but I still don’t get an acceptable network?

By the way, I am tracking a monkey. Tracking its limbs seems to be a big problem. In the tracking videos I got, the network often loses tracking of the limbs or confuses the left and right feet. This is also a problem I am facing that I hope to be optimized.

Regarding the interpolation of missing content you mentioned earlier, can you briefly talk about how I would do it?

Thanks again for your help! Thank you for your patience for a beginner！

**Comment by MMathisLab** [comment]
Hi Both, I'd ask we move this to the forum (https://forum.image.sc/tag/deeplabcut) , as this is not a git-hub issue; i.e. DLC is working. @ZiyiZhang0912 -- the plots above do show likely good tracking. Please create and watch the video; you need to set the p-cutoff >0.6 to handle occlusions.

Thanks to @KonradDanielewski for all his spot-on advice.
